heapy-imp/src/test/scala/MySuite.scala

```diff
@@ -0,0 +1,145 @@
+            )
+          )
+        )
+      )
+    )
+
+    // x := 1;
+    // counter := 0;
+    // while True do {
+    //     x := x + x;
+    //     counter := counter + 1;
+    //     if counter == 10 then break
+    // }
+    val i2 = build[ImpStmt](
+      // x := 1
+      build[ImpStmt](build[AExp]("x"), build[AExp](1)),
+      build[ImpStmt](
+        // counter := 0
+        build[ImpStmt](build[AExp]("counter"), build[AExp](0)),
+        // while True
+        build[ImpStmt](
+          BExp.True,
+          build[ImpStmt](
+            // x := x + x
+            build[ImpStmt](
+              build[AExp]("x"),
+              build[AExp](build[AExp]("x"), build[AExp]("x"))
+            ),
+            build[ImpStmt](
+              // counter := counter + 1
+              build[ImpStmt](
+                build[AExp]("counter"),
+                build[AExp](build[AExp]("counter"), build[AExp](1))
+              ),
+              // if 10 <= counter then break else skip
+              build[ImpStmt](
+                build[BExp](build[AExp](10), build[AExp]("counter")),
+                ImpStmt.Break,
+                ImpStmt.Skip
+              )
+            )
+          )
+        )
+      )
+    )
+
+    // x := 0x114514;
+    // h[x] = 1919810
+    val i3 = build[ImpStmt](
+      build[ImpStmt](build[AExp]("x"), build[AExp](0x114514)),
+      build[ImpStmt]("write", build[AExp]("x"), build[AExp](1919810))
+    )
+
+    // y := h[0x114514];
+    // y := y + 1;
+    // h[0x114514] = y
+    val i4 = build[ImpStmt](
+      build[ImpStmt](build[AExp]("y"), build[AExp](build[AExp](0x114514))),
+      build[ImpStmt](
+        // y := y + 1
+        build[ImpStmt](
+          build[AExp]("y"),
+          build[AExp](build[AExp]("y"), build[AExp](1))
+        ),
+        build[ImpStmt]("write", build[AExp](0x114514), build[AExp]("y"))
+      )
+    )
+
+    test("aeval basic test") {
+        val a1 = AExp.Nat(114514)
+        val a2 = AExp.Nat(1919810)
+        val a3 = AExp.Add(AddExp(a1, a2))
+        assertEquals(a3.aeval(Map()), 114514 + 1919810)
+    }
+
+    test("beval basic test") {
+        val a1 = AExp.Nat(114514)
+        val a2 = AExp.Nat(1919810)
+        val b  = BExp.Comp(CompExp(a1, a2))
+        assertEquals(b.beval(Map()), true)
+    }
+
+    test("eval basic test 1") {
+        val (sigma: ImpEvalContext, signal: Signal) = eval(i1)
+
+        assert(!signal.isBreak, "expect `signal` to be continue")
+        assertEquals(
+          sigma.size,
+          2,
+          "expect there are exact 2 variables in `sigma`"
+        )
+        assertEquals(sigma("x"), 11, "the final state of `x` should be 11")
+        assertEquals(sigma("y"), 55, "the final state of `y` should be 55")
+    }
+
+    test("eval basic test 2") {
+        val (sigma: ImpEvalContext, signal: Signal) = eval(i2)
+
+        assertEquals(sigma("counter"), 10)
+        assertEquals(sigma("x"), 1024)
+    }
+
+    test("heap basic test") {
```

Comment:
the test here shows how we seemingly "share" memory for two or even more applications [1], i.e., the latter can actually read the value from the previously modified memory region (by the former application(s)), just like any pointer ref/deref on heap (or global variable) - which we've seen a lot in C/C++.

[1] the application is just a "evaluable" `ImpStmt`, which could be evaluated by our built-in interpreter; interestingly, you could also see these `ImpStmt(s)` as "thread(s)" - since they share the same address space, and it's also possible to introduce "threading" (a.k.a. concurrency) for `heapy-imp` because we've already set up the global `heap`.